engine post (#871)
tobymao/sqlglot: posts/python_sql_engine.md

# Writing a Python SQL Engine from Scratch
[Toby Mao](https://www.linkedin.com/in/toby-mao/)

> ## Introduction

Reviewer: I think this is missing a summary of this post itself. That help readers know what to expect.

> When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half due to scope creep, I've ended up with a full fledge SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.

Reviewer 2: ```suggestion
When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half due to scope creep, I've ended up with a full fledged SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.
```

> When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half due to scope creep, I've ended up with a full fledge SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.

Reviewer: ```suggestion
When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half, I've ended up with a full-fledged SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.
```

Removed "due to scope creep" - I don't think it was necessary.

> I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder so that definitions could be reused across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my on SQL parser / transpiler.

Reviewer 2: ```suggestion
I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder, which enabled the reuse of definitions across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my own SQL parser / transpiler.
```

> I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder so that definitions could be reused across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my on SQL parser / transpiler.

Reviewer: I feel like this paragraph can go under "Why?", leaving the intro a little more concise.

Finally, the components that have been built to support execution can be used as a **foundation** for a faster engine. I'm inspired by what [Apache Calcite](https://github.com/apache/calcite) has done for the JVM world. Even though Python is commonly used for data, there hasn't been a Calcite for Python. So you could say that SQLGlot aims to be that framework. For example, it wouldn't take much work to replace the Python execution engine with numpy / pandas / arrow to become a respectably performing query engine. The implementation would be able to leverage the parser, optimizer, and logical planner, only needing to implement physical execution. There is a ton of work in the Python ecosystem around high performance vectorized computation which I think could benefit from a pure Python based [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) / [plan](https://en.wikipedia.org/wiki/Query_plan). Parsing and planning doesn't have to be fast when the bottleneck of running queries is processing terabytes of data. So having a Python based ecosystem around SQL is beneficial given the ease of development in Python despite not having bare metal performance.

> Parts of SQLGlot's toolkit are being used today by [Ibis](https://github.com/ibis-project/ibis), [Quokka](https://github.com/marsupialtail/quokka), and [Splink](https://github.com/moj-analytical-services/splink).

Reviewer: Might not be as exciting of a use case, but https://github.com/kelsin/mysql-mimic uses the executor as-is.

## How?

There are many steps involved with actually running a simple query like:

```sql
SELECT
  bar.a,
  b + 'y' AS b
FROM bar
JOIN baz
  ON bar.a = baz.a
WHERE bar.a > 1
```

In this post, I'll walk through all the steps SQLGlot takes to run this query over Python objects.

## Tokenizing

> The first step is to convert the sql string into a list of tokens. SQLGlot's tokenizer is quite simple and can be found [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py#L763). In a while loop, it checks each character and either appends the character to the current token or makes a new token.

Reviewer: The line where the Tokenizer starts is likely to change throughout the life of this post. We can link to a specific commit or remove the line number.

Running the SQLGlot tokenizer shows the output.

![Tokenizer Output](python_sql_engine_images/tokenizer.png)

> Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.

Reviewer 2: ```suggestion
Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that they can be preserved when transpiling SQL.
```

Reviewer: The sentence about comments doesn't feel necessary.

## Parsing

Once a SQL statement is tokenized, we don't need to worry about white space and other formatting, so it's easier to work with. We can now convert the list of tokens into an AST. The SQLGlot [parser](https://github.com/tobymao/sqlglot/blob/main/sqlglot/parser.py) is a hand written [recursive descent](https://en.wikipedia.org/wiki/Recursive_descent_parser) parser.

Similar to the tokenizer, it consumes the tokens sequentially but it uses a recursive algorithm instead. The tokens are converted into a single AST node which presents the SQL query. The SQLGlot parser was designed to support various dialects so it contains many options for overriding parsing functionality.

![Parser Output](python_sql_engine_images/parser.png)

The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.

## Optimizing

> Once we have our AST, we need to simplify it and put it into a consistent form to actually run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.

Reviewer 2: ```suggestion
Once we have our AST, we need to simplify it and transform it into a canonical form that enables us to run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.
```

> Once we have our AST, we need to simplify it and put it into a consistent form to actually run queries efficiently. For example, how do you know which table column 'b' comes from in `SELECT bar.a, b FROM bar JOIN baz`. Given a schema, we can infer that b comes from table baz. When optimizing queries, most engines first convert the AST into a logical plan and then optimize the plan. However, I chose to **optimize the AST directly** for two reasons.

Reviewer: While qualify_columns is super important to the optimizer, it's not really an optimization. I've seen this confuse people. I think we can just remove those two sentences. The paragraph will also flow better without it.

@GeorgeSittas - I think your change switches to passive voice. How about...

"Once we have our AST, we can transform it into an equivalent query that would produce the same results but execute for efficiently."

> 1. I wanted a way to generate 'canonical sql'. Given two SQL statements, I wanted to be able to identify if they are semantically equivalent. For example `select 2` and `select 1 + 1` are functionally the same even though the SQL query is the same. Having the optimizer output SQL would allow me to convert 1 + 1 into 2.

Reviewer: I don't think this really captures the _why_, which might confuse people.

I think the why has to do with versioning Tobiko models. But that's hard to sum up quickly...

Perhaps we should move this after "It's easier to debug", and say something about "canonical sql" being super useful in other use cases, separate from the executor. E.g. determining if two queries are semantically equivalent.

2. It's easier to debug and [validate](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/optimizer) the optimizations when the input and output are both SQL.

I've yet to find another engine that takes this approach, but I'm quite happy with this decision. The optimizer currently does not perform any "physical optimizations" like join reordering. Those are left to the execution layer as additional statistics and information could become relevant.

![Optimizer Output](python_sql_engine_images/optimizer.png)

The optimizer currently has [17 rules](https://github.com/tobymao/sqlglot/tree/main/sqlglot/optimizer). Each of these rules is applied, transforming the AST in place. The combination of these rules creates "canonical" sql which can then be more easily converted into a logical plan and then executed.

Some example rules are:

### qualify\_tables and qualify_columns

> - Adds all db/catalog qualifiers to tables and forces an alias.
> - Ensure each column is unambiguous and expand stars.

Reviewer 2: ```suggestion
- Adds all db/catalog qualifiers to tables and forces aliases.
- Ensures each column is unambiguous and expands star projections.
```

> ```sql

Reviewer: We have a mix of inline code blocks and images.

Where do you plan to post this? Will it support inline code blocks with syntax highlighting?

Author: i plan to leave it in the mark down

> SELECT * FROM x`;

Reviewer: ```suggestion
SELECT * FROM x;
```

> SELECT "db"."x" AS "x";

Reviewer: ```suggestion
SELECT x.a AS a, x.b AS b FROM db.x AS x;
```

```

### simplify

> Boolean and math simplification. Check out all the [test cases](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/optimizer/simplify.sql).

Reviewer 2: ```suggestion
Simplifies math expressions. Check out all the [test cases](https://github.com/tobymao/sqlglot/blob/main/tests/fixtures/optimizer/simplify.sql).
```

```sql
((NOT FALSE) AND (x = x)) AND (TRUE OR 1 <> 3);
x = x;

1 + 1;
2;
```

### normalize
Attempts to convert all predicates into [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form).

```sql
-- DNF
(A AND B) OR (B AND C AND D);

-- CNF
(A OR C) AND (A OR D) AND B;
```

### unnest\_subqueries
Converts subqueries in predicates into joins.

```sql
-- The subquery can be converted into a left join
SELECT *
FROM x AS x
WHERE (
  SELECT y.a AS a
  FROM y AS y
  WHERE x.a = y.a
) = 1;

SELECT *
FROM x AS x
LEFT JOIN (
  SELECT y.a AS a
  FROM y AS y
  WHERE TRUE
  GROUP BY y.a
) AS "_u_0"
  ON x.a = "_u_0".a
WHERE ("_u_0".a = 1 AND NOT "_u_0".a IS NULL)'
```

### pushdown_predicates

> Push down filters into the innermost query.

Reviewer 2: ```suggestion
Pushes down filters into the innermost query.
```

```sql
SELECT *
FROM (
  SELECT *
  FROM x AS x
) AS y
WHERE y.a = 1;

SELECT *
FROM (
  SELECT *
  FROM x AS x
  WHERE y.a = 1
) AS y WHERE TRUE
```

### annotate_types

> Infer all types throughout the AST given schema information and function type definitions.

Reviewer 2: ```suggestion
Infers all types throughout the AST given schema information and function type definitions.
```

## Planning
After the SQL AST has been "optimized", it's much easier to [convert into a logical plan](https://github.com/tobymao/sqlglot/blob/main/sqlglot/planner.py). The AST is traversed and converted into a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) consisting of one of five steps. The different types of steps are:

### Scan

> The scan step selects columns from a table, applies projections, and finally filters the table.

Reviewer: I think we can just remove "The _ step" from all these

# Writing a Python SQL Engine from Scratch
[Toby Mao](https://www.linkedin.com/in/toby-mao/)

## Introduction

When I first started writing SQLGlot in early 2021, my goal was just to translate SQL queries from SparkSQL to Presto and vice versa. However, over the last year and a half, I've ended up with a full-fledged SQL engine. SQLGlot can now parse and transpile between [18 SQL dialects](https://github.com/tobymao/sqlglot/blob/main/sqlglot/dialects/__init__.py) and can execute all 24 [TPC-H](https://www.tpc.org/tpch/) SQL queries. The parser and engine are all handwritten from scratch with Python.

This post will cover [why](#why) I went through the effort of creating a Python SQL engine and [how](#how) a simple query goes from a string to actually transforming data. The following steps are briefly summarized:

* [Tokenizing](#tokenizing)
* [Parsing](#parsing)
* [Optimizing](#optimizing)
* [Planning](#planning)
* [Executing](#executing)

## Why?

> I started working on SQLGlot because of my work on the [experimentation and metrics platform](https://netflixtechblog.com/reimagining-experimentation-analysis-at-netflix-71356393af21) at Netflix, where I built tools that allowed data scientists to define and compute SQL based metrics. Netflix relied on multiple engines to query data (Spark, Presto, and Druid) so my team built the metrics platform around [PyPika](https://github.com/kayak/pypika), a Python SQL query builder so that definitions could be reused across multiple engines. However, it became quickly apparent that writing python code to programatically generate SQL was challenging for data scientists, especially those with academic backgrounds since they were mostly familiar with R and SQL. But at the time, the only Python SQL parser was [sqlparse]([https://github.com/andialbrecht/sqlparse) which is not actually a parser, but a tokenizer, so having users write raw SQL into the platform wasn't really an option. Some time later, I randomly stumbled across [Crafting Interpreters](https://craftinginterpreters.com/) and realized that I could use it as a guide towards creating my on SQL parser / transpiler.

Reviewer 3: > creating my on SQL parser / transpiler.

typo - "my own".

Why did I do this? Isn't a Python SQL engine going to be extremely slow?

The main reason why I ended up building a SQL engine was ... just for **entertainment**. It's been fun learning about all the things required to actually run a SQL query and seeing it actually work is extremely rewarding. Before SQLGlot, I had zero experience with lexers, parsers, or compilers.

> In terms of practical use cases, I planned to use the Python SQL engine for unit testing SQL pipelines. Big data pipelines are tough to test because many of the engines are not open source and cannot be run locally. With SQLGlot, you can take a SQL query targeting a warehouse like [Snowflake](https://www.snowflake.com/en/), and seamlessly run it in CI on mock Python data. It's easy to mock data and create arbitrary [UDFs](https://en.wikipedia.org/wiki/User-defined_function) because everything is just Python.  Although the implementation is slow and unsuitable for large amounts of data (> 1 millon rows), there's very little overhead / startup and can run queries on test data in a couple of milliseconds.

Reviewer 3: > and can run queries on test data in a couple of milliseconds.

and "it" can run? Otherwise the sentence feels a little off.

## How?

There are many steps involved with actually running a simple query like:

```sql
SELECT
  bar.a,
  b + 'y' AS b
FROM bar
JOIN baz
  ON bar.a = baz.a
WHERE bar.a > 1
```

In this post, I'll walk through all the steps SQLGlot takes to run this query over Python objects.

## Tokenizing

The first step is to convert the sql string into a list of tokens. SQLGlot's tokenizer is quite simple and can be found [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py). In a while loop, it checks each character and either appends the character to the current token or makes a new token.

Running the SQLGlot tokenizer shows the output.

![Tokenizer Output](python_sql_engine_images/tokenizer.png)

Each keyword has been converted to a SQLGlot Token object. Each token has some metadata associated with it like line / column information for error messages. Comments are also a part of the token so that comments can be preserved.

## Parsing

> Once a SQL statement is tokenized, we don't need to worry about white space and other formatting, so it's easier to work with. We can now convert the list of tokens into an AST. The SQLGlot [parser](https://github.com/tobymao/sqlglot/blob/main/sqlglot/parser.py) is a hand written [recursive descent](https://en.wikipedia.org/wiki/Recursive_descent_parser) parser.

Reviewer 3: > to worry about white space

white space(s) - plural? 

> the list of tokens into an AST

worth expanding AST in parenthesis (Abstract Syntax Tree) the first time you refer to it. Perhaps add a link to Wikipedia.

Author: it's done earlier

Similar to the tokenizer, it consumes the tokens sequentially but it uses a recursive algorithm instead. The tokens are converted into a single AST node which presents the SQL query. The SQLGlot parser was designed to support various dialects so it contains many options for overriding parsing functionality.

![Parser Output](python_sql_engine_images/parser.png)

> The AST is a generic representation of a given SQL query. Each dialect can override or implement its own generator that can convert an AST object into syntatically correct SQL.

Reviewer 3: > into syntatically correct SQL.

"into a syntatically correct SQL expression."?
